File: Vue/src/utils/ColorUtils.ts
export interface RGBColor {
	r: number;
	g: number;
	b: number;
	a?: number;
}

export function rgbToHsv(r: number, g: number, b: number): { h: number; s: number; v: number } {
	const max = Math.max(r, g, b);
	const min = Math.min(r, g, b);
	const delta = max - min;

	let h = 0;
	if (delta !== 0) {
		if (max === r) {
			h = 60 * (((g - b) / delta) % 6);
		} else if (max === g) {
			h = 60 * ((b - r) / delta + 2);
		} else {
			h = 60 * ((r - g) / delta + 4);
		}
	}

	if (h < 0) h += 360;

	const s = max === 0 ? 0 : delta / max;
	const v = max;

	return { h, s, v };
}

export function hsvToRgb(h: number, s: number, v: number): RGBColor {
	const c = v * s;
	const x = c * (1 - Math.abs(((h / 60) % 2) - 1));
	const m = v - c;

	let r1 = 0,
		g1 = 0,
		b1 = 0;

	if (h < 60) {
		r1 = c;
		g1 = x;
		b1 = 0;
	} else if (h < 120) {
		r1 = x;
		g1 = c;
		b1 = 0;
	} else if (h < 180) {
		r1 = 0;
		g1 = c;
		b1 = x;
	} else if (h < 240) {
		r1 = 0;
		g1 = x;
		b1 = c;
	} else if (h < 300) {
		r1 = x;
		g1 = 0;
		b1 = c;
	} else {
		r1 = c;
		g1 = 0;
		b1 = x;
	}

	return {
		r: r1 + m,
		g: g1 + m,
		b: b1 + m,
	};
}

export function deepen(color: RGBColor, saturationBoost: number = 0.4, darknessFactor: number = 0.7): RGBColor {
	const hsv = rgbToHsv(color.r, color.g, color.b);

	hsv.s = Math.min(1.0, hsv.s + saturationBoost);

	hsv.v = hsv.v * darknessFactor;

	const rgb = hsvToRgb(hsv.h, hsv.s, hsv.v);

	return {
		r: rgb.r,
		g: rgb.g,
		b: rgb.b,
		a: color.a,
	};
}

export function linearInterpolate(result: number[], start: number[], end: number[], t: number): void {
	for (let i = 0; i < start.length; i++) {
		const startVal = start[i];
		const endVal = end[i];
		if (startVal !== undefined && endVal !== undefined) {
			result[i] = startVal + (endVal - startVal) * t;
		}
	}
}

export function floatArrayToRGBColor(floatArray: number[], index: number): RGBColor {
	const offset = index * 4;
	return {
		r: floatArray[offset] ?? 0,
		g: floatArray[offset + 1] ?? 0,
		b: floatArray[offset + 2] ?? 0,
		a: floatArray[offset + 3] ?? 1,
	};
}

export function rgbToCSS(color: RGBColor): string {
	const r = Math.round(color.r * 255);
	const g = Math.round(color.g * 255);
	const b = Math.round(color.b * 255);
	const a = color.a ?? 1;
	return `rgba(${r}, ${g}, ${b}, ${a})`;
}

export function floatArrayToCSS(floatArray: number[], index: number): string {
	const offset = index * 4;
	const r = Math.round((floatArray[offset] ?? 0) * 255);
	const g = Math.round((floatArray[offset + 1] ?? 0) * 255);
	const b = Math.round((floatArray[offset + 2] ?? 0) * 255);
	const a = floatArray[offset + 3] ?? 1;
	return `rgba(${r}, ${g}, ${b}, ${a})`;
}

File: KMP_Js/composeApp/src/jsMain/kotlin/plus/yumeyuka/yumeyuka/utils/ColorUtils.kt
package plus.yumeyuka.yumeyuka.utils

import androidx.compose.ui.graphics.Color
import kotlin.math.max
import kotlin.math.min

private fun rgbToHsv(r: Float, g: Float, b: Float): FloatArray {
    val max = max(r, max(g, b))
    val min = min(r, min(g, b))
    val delta = max - min
    
    val h = when {
        delta == 0f -> 0f
        max == r -> 60f * (((g - b) / delta) % 6f)
        max == g -> 60f * (((b - r) / delta) + 2f)
        else -> 60f * (((r - g) / delta) + 4f)
    }.let { if (it < 0f) it + 360f else it }
    
    val s = if (max == 0f) 0f else delta / max
    
    val v = max
    
    return floatArrayOf(h, s, v)
}

private fun hsvToRgb(h: Float, s: Float, v: Float): FloatArray {
    val c = v * s
    val x = c * (1f - kotlin.math.abs((h / 60f) % 2f - 1f))
    val m = v - c
    
    val (r1, g1, b1) = when {
        h < 60f -> Triple(c, x, 0f)
        h < 120f -> Triple(x, c, 0f)
        h < 180f -> Triple(0f, c, x)
        h < 240f -> Triple(0f, x, c)
        h < 300f -> Triple(x, 0f, c)
        else -> Triple(c, 0f, x)
    }
    
    return floatArrayOf(r1 + m, g1 + m, b1 + m)
}

fun Color.deepen(saturationBoost: Float = 0.4f, darknessFactor: Float = 0.7f): Color {
    val hsv = rgbToHsv(red, green, blue)
    
    hsv[1] = min(1.0f, hsv[1] + saturationBoost)
    
    hsv[2] = hsv[2] * darknessFactor
    
    val rgb = hsvToRgb(hsv[0], hsv[1], hsv[2])
    return Color(
        red = rgb[0],
        green = rgb[1],
        blue = rgb[2],
        alpha = alpha
    )
}

fun List<Color>.deepenAll(saturationBoost: Float = 0.4f, darknessFactor: Float = 0.7f): List<Color> {
    return map { it.deepen(saturationBoost, darknessFactor) }
}